mm2src/mm2_metamask/src/metamask_provider.rs

```diff
@@ -0,0 +1,246 @@
+use crate::eth_provider::EthProvider;
+use crate::metamask_error::{MetamaskError, MetamaskResult};
+use futures::lock::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use itertools::Itertools;
+use mm2_err_handle::prelude::*;
+use serde::Serialize;
+use serde_derive::Deserialize;
+use serde_json::Value as Json;
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::Arc;
+
+/// `MetamaskProvider` is designed the way that there can be only one active session at the moment.
+/// This is highly unlikely that the channel will be full with `capacity = 1024` during this session.
+const ETH_COMMAND_CHANNEL_CAPACITY: usize = 1024;
+const EIP712_DOMAIN: &str = "EIP712Domain";
+
+macro_rules! eth_rpc_await {
+    ($selff:ident, $method:expr $(, $arg_name:expr)*) => {{
+        let params = vec![
+            $(
+                serde_json::value::to_value($arg_name)
+                    .map_to_mm(|e| MetamaskError::ErrorSerializingArguments(e.to_string()))?
+            ),*
+        ];
+        $selff
+            .eth_provider
+            .invoke_method($method.to_string(), params)
+            .await
+            .mm_err(MetamaskError::from)
+    }}
+}
+
+#[derive(Clone)]
+pub struct MetamaskProvider {
+    eth_provider: Arc<AsyncMutex<EthProvider>>,
+}
+
+impl MetamaskProvider {
+    pub fn detect_metamask_provider() -> MetamaskResult<MetamaskProvider> {
+        let eth_provider = EthProvider::detect_ethereum_provider(ETH_COMMAND_CHANNEL_CAPACITY)
+            .or_mm_err(|| MetamaskError::EthProviderNotFound)?;
+        Ok(MetamaskProvider {
+            eth_provider: Arc::new(AsyncMutex::new(eth_provider)),
+        })
+    }
+
+    /// Creates a session that can be used to invoke methods.
+    /// We need to limit the number of concurrent requests to one.
```

Comment:
Could you please provide more details on why we `need to limit the number of concurrent requests to one`?

Reply:
Sure! The first reason is switching `chainId` and further withdrawing:
```rust
// Thread 1 (witdhraw)
let metamask = metamask_provider();
metamask.wallet_switch_ethereum_chain(BNB_CHAIN_ID).await?; // thread1_1
metamask.eth_send_transaction(tx).await?; // thread1_2

// Thread 2 (swap)
let metamask = metamask_provider();
metamask.wallet_switch_ethereum_chain(ETH_CHAIN_ID).await?; // thread2_1
metamask.eth_send_transaction(tx).await?; // thread2_2
```
It may happen in the following order:
1. thread1_1
2.thread2_1 
3. thread2_2
4. thread1_2 (fail)

Another reason is that I use a bounded `mpsc` channel that takes `&mut self` to send any command:
https://github.com/KomodoPlatform/atomicDEX-API/blob/cb4aabadd95ad6fa9d3cd01385f1f96b5558387c/mm2src/mm2_metamask/src/eth_provider.rs#L44-L56

So if there is a need to send a single RPC request (for example, my_balance), the `MetamaskSession<'a>` instance can be dropped immediately after the RPC call. But if it's needed to perform multiple interactions at once, the session should be held.

Reply:
I see, thanks for clarification!